### PR TITLE
[CPU] Dot product attention : CPU fused attention custom call

### PR DESCRIPTION
**Pull request by sachinmuradi**
This PR is to get a initial feedback on the feature.

This PR adds implementation for existing dot_product_attention op in Jax, named as 'cpu_fused_attention' implementation for CPU devices.
This provides a modified lowering of the dot product attention via MLIR custom call and hlo operations. The custom call can be eventually lowered to any backend library calls.

`mlir.custom_call(
      "custom_call_target_name",
      result_types,
      operands,
      operand_layouts,
      backend_config,
      result_layouts
    )`

Currently custom call lowering is implemented for OneDnn CPU backend in XLA, but can be extended to support other library backend that supports fused attention kernel ,for ex XNNPACK/YNNPACK.

Currently the feature is restricted to : 
1) Linux OS
2) FP32, BF16, FP16 datatypes
3) Only activated when XNNPACK is not enabled in XLA (--xla_cpu_use_xnnpack=false)
4) Only activated when OneDnn custom call is used with XLA (--xla_cpu_experimental_onednn_custom_call=true)
5) For the draft, we restricted it to Intel CPU platforms, as the validation is done only on Intel CPUs as of now (But can be scaled to other CPU architecture too)
6) Limited attention types are supported ( No support for bias / sliding window attention / return residue )

If any of the conditions above not met , the implementation will fallback to 'xla' implementation.

### CODE REVIEW

**Comment by sachinmuradi** [comment]
Hi @skye , Would you or someone be able to review the PR? Thank you

File: tests/nn_test.py [hunk 1]
```diff
@@ -103,6 +104,10 @@ def create_mxfp8_configs_if_available():
 
   return [nn.get_scaled_dot_general_config("mxfp8") for _ in range(3)]
 
+def _check_onednn_backend(fn, *args, **kwargs):
```
Comment (bchetioui) [inline tests/nn_test.py:107]: If this is only used in your new test, how about inlining this?

**Comment by skye** [comment]
@bchetioui and perhaps @hawkinsp can take a look! (Peter is out so may not be able to catch up in time)

File: jax/_src/cpu_lib_ops/fused_attention_cpu.py [hunk 2]
```diff
@@ -0,0 +1,390 @@
...
+      return False
+
+def can_use_cpu_fused_attention(dtype):
+    xla_flags = os.environ.get("XLA_FLAGS", "")
+    flags = xla_flags.split()
+    # XLA sets xla_cpu_use_xnnpack flag to true by default
+    # Currently fused attention op is only supported with oneDNN path
+    # so we need to ensure that xnnpack path is disabled
+    # This can be scaled to support xnnpack path in future if needed
+    if ("--xla_cpu_use_xnnpack=false" not in flags):
```
Comment (bchetioui) [inline jax/_src/cpu_lib_ops/fused_attention_cpu.py:54]: Instead of a negative check, is there a way to turn this into a positive check? As is, it seems like XLA could choose to pick yet a different path from XNNPACK and oneDNN?

E.g., I see that XLA has a `--xla_cpu_use_onednn` knob. Would checking that work?

File: jax/_src/cpu_lib_ops/fused_attention_cpu.py [hunk 3]
```diff
@@ -0,0 +1,390 @@
...
+from jax._src.lib.mlir import ir
+from jax._src.lib.mlir.dialects import hlo
+
+logger = logging.getLogger(__name__)
+
+def _is_dtype_supported(dtype: jnp.dtype, cpuinfo: str):
+   if dtype == jnp.float32:
+      return True
+   elif dtype == jnp.bfloat16:
+      has_bf16 = any(flag in cpuinfo for flag in ["avx512f", "avx_ne_convert", "amx_bf16"])
```
Comment (bchetioui) [inline jax/_src/cpu_lib_ops/fused_attention_cpu.py:38]: Should this be `return any(flag in cpuinfo for flag in ["avx512f", "avx_ne_convert", "amx_bf16"])`?

File: jax/_src/cpu_lib_ops/fused_attention_cpu.py [hunk 4]
```diff
@@ -0,0 +1,390 @@
...
+      return False
+
+    if ("--xla_cpu_experimental_onednn_custom_call=true" not in flags):
+      logger.info("Not using CPU Fused Attention, to use add --xla_cpu_experimental_onednn_custom_call=true to XLA_FLAGS")
+      return False
+    # Only supported for linux
+    if platform.system().lower() != "linux":
+      logger.info("CPU Fused Attention is supported only for linux OS")
+      return False
+    # Currentl supported for CPU backend only
```
Comment (bchetioui) [inline jax/_src/cpu_lib_ops/fused_attention_cpu.py:65]: Currently

File: jax/_src/nn/functions.py [hunk 5]
```diff
@@ -1226,6 +1235,42 @@ def _check_shape_and_dtype(t: Array | None, shape: Sequence[int],
...
         out = (out, residual)
+    case 'cpu_fused_attention':
+      use_cpu_fusedop = True
+
+      if bias is not None:
+        warnings.warn("CPU fused attention doesn't support bias, falling back to xla implementation")
+        use_cpu_fusedop = False
+
+      if query_seq_lengths is not None or key_value_seq_lengths is not None:
+        warnings.warn("CPU fused attention doesn't support query_seq_lengths / key_value_seq_lengths, falling back to xla implementation")
```
Comment (bchetioui) [inline jax/_src/nn/functions.py:1246]: nit: here and below, let's capitalize "XLA" in the warnings

File: jax/_src/nn/functions.py [hunk 6]
```diff
@@ -1226,6 +1235,42 @@ def _check_shape_and_dtype(t: Array | None, shape: Sequence[int],
...
+        use_cpu_fusedop = False
+
+      if return_residual:
+        warnings.warn("CPU fused attention doesn't support return_residual, falling back to xla implementation")
+        use_cpu_fusedop = False
+
+      # We also check cpuinfo to make sure we use cpu_fused_attention implementation
+      # only for certain cpu features
+      if not (can_use_cpu_fused_attention(query_arr.dtype) and use_cpu_fusedop):
+        warnings.warn("Dot product attention not compatible with cpu_fused_attention, falling back to xla implementation")
```
Comment (bchetioui) [inline jax/_src/nn/functions.py:1260]: These warnings can stack up if several properties are missing, making the error messages a bit redundant.

I think we can fix this (and simplify the implementation) by just raising on the first missing property we encounter, with no fallback. After all, the user requested that we use `'cpu_fused_attention'`---so if we can't, it should likely be an error.

File: jax/_src/nn/functions.py [hunk 7]
```diff
@@ -1051,10 +1055,11 @@ def dot_product_attention(
     *,
     scale: float | None = None,
     is_causal: bool = False,
+    attn_logits_soft_cap: Any | None = None,
     query_seq_lengths: ArrayLike | None = None,
     key_value_seq_lengths: ArrayLike | None = None,
     local_window_size: int | tuple[int, int] | None = None,
-    implementation: Literal['xla', 'cudnn'] | None = None,
+    implementation: Literal['xla', 'cudnn', 'cpu_fused_attention'] | None = None,
```
Comment (bchetioui) [inline jax/_src/nn/functions.py:1062]: Please update the docstring accordingly.

Also, should we call this `onednn` to be specific wrt which implementation we target?

File: tests/nn_test.py [hunk 8]
```diff
@@ -380,6 +385,50 @@ def bwd_ans(x, bias, mask):
...
+  @parameterized.named_parameters([
+      dict(
+          testcase_name=f"dtype_{dtype.__name__}_mode_{mode}_batch_{batch}_group_{group}_qLen_{q_len}",
+          dtype=dtype,
+          causal_mode=mode,
+          group_num=group,
+          query_seq_len=q_len,
+          batch_size=batch,
+      )
+      for dtype in [jnp.float32, jnp.bfloat16]
```
Comment (bchetioui) [inline tests/nn_test.py:398]: You also have a branch testing specifically for `jnp.float16`---let's add this data type to the test as well.

File: jax/_src/cpu_lib_ops/fused_attention_cpu.py [hunk 9]
```diff
@@ -0,0 +1,390 @@
...
+
+def can_use_cpu_fused_attention(dtype):
+    xla_flags = os.environ.get("XLA_FLAGS", "")
+    flags = xla_flags.split()
+    # XLA sets xla_cpu_use_xnnpack flag to true by default
+    # Currently fused attention op is only supported with oneDNN path
+    # so we need to ensure that xnnpack path is disabled
+    # This can be scaled to support xnnpack path in future if needed
+    if ("--xla_cpu_use_xnnpack=false" not in flags):
+      logger.info("Not using CPU Fused Attention, to use set XLA_FLAGS=--xla_cpu_use_xnnpack=false")
```
Comment (bchetioui) [inline jax/_src/cpu_lib_ops/fused_attention_cpu.py:55]: I don't think we want all this logging every time we check whether we can use `cpu_fused_attention`. Either remove it, or alternatively use `logging.vlog(1, msg)` (or some higher level) in order to allow the user to enable it at will, but disable it by default.

(Same holds for all the `logger.info` calls in this function.)

File: jax/_src/cpu_lib_ops/fused_attention_cpu.py [hunk 10]
```diff
@@ -0,0 +1,390 @@
+# Copyright 2025 The JAX Authors.
```
Comment (bchetioui) [inline jax/_src/cpu_lib_ops/fused_attention_cpu.py:1]: Since most of the logic here is specialized to oneDNN, I suggest we follow the directory structure we used for `cudnn`, and call the folder `onednn`

File: jax/_src/cpu_lib_ops/fused_attention_cpu.py [hunk 11]
```diff
@@ -0,0 +1,390 @@
...
+      has_bf16 = any(flag in cpuinfo for flag in ["avx512f", "avx_ne_convert", "amx_bf16"])
+      return True
+   elif dtype == jnp.float16:
+      has_f16 = any(flag in cpuinfo for flag in ["avx_ne_convert", "amx_fp16"])
+      has_f16_combined_flags = "avx512_fp16" in cpuinfo and "avx512bw" in cpuinfo
+      return has_f16 or has_f16_combined_flags
+   else:
+      return False
+
+def can_use_cpu_fused_attention(dtype):
```
Comment (bchetioui) [inline jax/_src/cpu_lib_ops/fused_attention_cpu.py:47]: Here and in most places, let's replace `cpu` with `onednn`!

File: jax/_src/cpu_lib_ops/fused_attention_cpu.py [hunk 12]
```diff
@@ -0,0 +1,390 @@
...
+        is_CPU_intel = "intel" in cpuinfo
+    except Exception as e:
+        logger.info(f"CPU Fused Attention Failed to detect CPU info: {e}", file=sys.stderr)
+        return False
+
+    is_supported_dtype = _is_dtype_supported(dtype, cpuinfo)
+
+    return is_CPU_intel and is_supported_dtype
+
+def _check_layout(query, key, value):
```
Comment (bchetioui) [inline jax/_src/cpu_lib_ops/fused_attention_cpu.py:90]: I assume you intended to call this in the abstract eval rule for your primitive, but this is missing at the moment.

File: jax/_src/cpu_lib_ops/fused_attention_cpu.py [hunk 13]
```diff
@@ -0,0 +1,390 @@
...
+  vB, vS, vN, vH = value.shape
+
+  check_eq(qB, kB, vB, "QKV batch")
+  check_eq(qH, kH, vH, "QKV dim_per_head")
+  if kN != vN:
+    raise ValueError(f"KV must have same number of heads, got {kN} vs {vN}")
+  if kS != vS:
+    raise ValueError(f"KV must have same seq length, got {kS} vs {vS}")
+  if qN % kN != 0:
+     raise ValueError(f"Number of Query heads {qN} must be multiple of KV heads {kN}")
```
Comment (bchetioui) [inline jax/_src/cpu_lib_ops/fused_attention_cpu.py:117]: nit: "Q heads"

File: jax/_src/cpu_lib_ops/fused_attention_cpu.py [hunk 14]
```diff
@@ -0,0 +1,390 @@
...
+
+def _get_causal_mask(T, S):
+  mask = jnp.tril(jnp.ones((T, S), dtype=jnp.bool_))
+  return mask[None, None, :, :]
+
+def _reshape_to_grouped(t, num_query_heads, group_size, num_key_heads):
```
Comment (bchetioui) [inline jax/_src/cpu_lib_ops/fused_attention_cpu.py:123]: Maybe inline this? It's not clear what properties `t` is supposed to uphold from looking at this code, and it is only used in a single place. The logic will be more parsable if this is inlined at the use site.

File: jax/_src/cpu_lib_ops/fused_attention_cpu.py [hunk 15]
```diff
@@ -0,0 +1,390 @@
...
+  check_eq(qB, kB, vB, "QKV batch")
+  check_eq(qH, kH, vH, "QKV dim_per_head")
+  if kN != vN:
+    raise ValueError(f"KV must have same number of heads, got {kN} vs {vN}")
+  if kS != vS:
+    raise ValueError(f"KV must have same seq length, got {kS} vs {vS}")
+  if qN % kN != 0:
+     raise ValueError(f"Number of Query heads {qN} must be multiple of KV heads {kN}")
+
+def _get_causal_mask(T, S):
```
Comment (bchetioui) [inline jax/_src/cpu_lib_ops/fused_attention_cpu.py:119]: This is really short and called only once, let's inline it

File: jax/_src/cpu_lib_ops/fused_attention_cpu.py [hunk 16]
```diff
@@ -0,0 +1,390 @@
...
+
+  tB, tN, tT, tS = t.shape
+  if tN == 1:
+    t = jnp.broadcast_to(t[:, :, None, :, :], (tB, tN, group_size, tT, tS))
+  else:
+    assert tN == num_query_heads
+    t = jnp.reshape(t, (tB, num_key_heads, group_size, tT, tS))
+  return t
+
+_custom_call_name_map = {
```
Comment (bchetioui) [inline jax/_src/cpu_lib_ops/fused_attention_cpu.py:133]: Please add a comment explaining what this map contains, and especially what are the semantics of the keys

File: jax/_src/cpu_lib_ops/fused_attention_cpu.py [hunk 17]
```diff
@@ -0,0 +1,390 @@
...
+    (False, False) : "QK_SOFTMAX_AV",
+    (True, False) : "QK_SELECT_SOFTMAX_AV",
+    (True, True): "QK_CAPPED_SELECT_SOFTMAX_AV",
+}
+
+def _get_custom_call_name(is_masked, is_attn_soft_cap):
+  return _custom_call_name_map[(is_masked, is_attn_soft_cap)]
```
Comment (bchetioui) [inline jax/_src/cpu_lib_ops/fused_attention_cpu.py:140]: Let's inline this lookup

File: jax/_src/cpu_lib_ops/fused_attention_cpu.py [hunk 18]
```diff
@@ -0,0 +1,390 @@
...
+
+def dot_product_attention(
+    query: Array,
+    key: Array,
+    value: Array,
+    mask: Array | None,
+    is_causal: bool,
+    scale: float,
+    attn_logits_soft_cap: Any | None = None):
+    r"""CPU implementation of Fused Scaled dot product attention function.
```
Comment (bchetioui) [inline jax/_src/cpu_lib_ops/fused_attention_cpu.py:323]: ~~CPU~~OneDNN

File: jax/_src/cpu_lib_ops/fused_attention_cpu.py [hunk 19]
```diff
@@ -0,0 +1,390 @@
...
+      is_causal: If true, causal attention will be applied.
+
+    Returns:
+      An array of the attention output with the same shape as :code:`query`.
+    """
+
+    # query shape: B, T, N, H
+    # key/value shape: B, S, K, H
+    is_gqa = False
+    if query.shape[2] != key.shape[2]:
```
Comment (bchetioui) [inline jax/_src/cpu_lib_ops/fused_attention_cpu.py:361]: You can write this more concisely as `is_gqa = query.shape[2] != key.shape[2]`?

File: jax/_src/cpu_lib_ops/fused_attention_cpu.py [hunk 20]
```diff
@@ -0,0 +1,390 @@
...
+        is_gqa = True
+    S = key.shape[1]
+    K = key.shape[2]
+    B, T, N, H = query.shape
+    G = 1
+
+    if is_gqa:
+        G = N // K
```
Comment (bchetioui) [inline jax/_src/cpu_lib_ops/fused_attention_cpu.py:369]: Note, you can always do `G = N // K`; when this is not the GQA case, the division will return `1`.

File: jax/_src/cpu_lib_ops/fused_attention_cpu.py [hunk 21]
```diff
@@ -0,0 +1,390 @@
...
+      An array of the attention output with the same shape as :code:`query`.
+    """
+
+    # query shape: B, T, N, H
+    # key/value shape: B, S, K, H
+    is_gqa = False
+    if query.shape[2] != key.shape[2]:
+        is_gqa = True
+    S = key.shape[1]
+    K = key.shape[2]
```
Comment (bchetioui) [inline jax/_src/cpu_lib_ops/fused_attention_cpu.py:364]: We should raise if `K` does not divide `N`

File: tests/nn_test.py [hunk 22]
```diff
@@ -380,6 +385,50 @@ def bwd_ans(x, bias, mask):
...
+    if not can_use_cpu_fused_attention(dtype):
+      raise unittest.SkipTest("Skipping as Dot product attention is not compatible.")
+    sdpa = nn.dot_product_attention
+    B, S, T, N, H, G = batch_size, 256, query_seq_len, 8, 256, group_num
+    keys = random.split(random.PRNGKey(0), 4)
+    Q = random.normal(keys[0], (B, T, N, H), dtype)
+    K = random.normal(keys[1], (B, S, N // G, H), dtype)
+    V = random.normal(keys[2], (B, S, N // G, H), dtype)
+    causal_mask = None
+    is_causal = causal_mode == 'is_causal'
```
Comment (bchetioui) [inline tests/nn_test.py:414]: It seems like you support `is_causal` with another mask being specified. In that case, can we test that configuration as well?

**Review by bchetioui — CHANGES_REQUESTED** [review]
Apologies for the delay---here is a first review!
